packages/manager/src/features/CloudPulse/Utils/CloudPulseWidgetUtils.ts

```diff
@@ -103,3 +140,75 @@ export const getCloudPulseMetricRequest = (
           },
   };
 };
+
+const getLabelName = (
+  labelObject: {
+    flags: FlagSet;
+    label: string;
+    serviceType: string;
+    unit: string;
+  },
+  metric: { [lable: string]: string },
+  resources: CloudPulseResources[]
+): string => {
+  // aggregated metric, where metric keys will be 0
+  if (!Object.keys(metric).length) {
+    // in this case return widget label and unit
+    return labelObject.label + ' (' + labelObject.unit + ')';
+  }
+
+  const results =
+    labelObject.flags?.aclpResourceTypeMap?.filter(
+      (obj: CloudPulseResourceTypeMapFlag) =>
+        obj.serviceType === labelObject.serviceType
+    ) ?? [];
+
+  const flag = results?.length ? results[0] : undefined;
+
+  return getDimensionName(metric, flag, resources);
+};
+
+export const getDimensionName = (
+  metric: { [label: string]: string },
+  flag: CloudPulseResourceTypeMapFlag | undefined,
+  resources: CloudPulseResources[]
+): string => {
+  let labelName = '';
+  Object.keys(metric).forEach((key) => {
+    if (key === flag?.dimensionKey) {
+      const mappedName = mapResourceIdToName(
+        metric[flag.dimensionKey],
+        resources
+      );
+
+      if (mappedName?.length) {
+        labelName +=
+          mapResourceIdToName(metric[flag.dimensionKey], resources) + '_';
+      }
+    } else if (metric[key]) {
+      labelName += metric[key] + '_';
+    }
+  });
+
+  return labelName.slice(0, -1);
+};
+
+export const mapResourceIdToName = (
+  id: string,
+  resources: CloudPulseResources[]
+): string => {
+  if (!resources?.length) {
+    return id;
+  }
+
+  if (!id) {
+    return '';
+  }
+
+  const index = resources.findIndex(
+    (resourceObj) => resourceObj.id && resourceObj.id === id
+  );
+
+  // if we are able to find a match, then return the label, else return id
+  return index != -1 ? resources[index].label : id;
```

Comment:
Nit: You're choice to add or not - Nothing wrong with approach, just could slightly readability 👍 
```suggestion
  const resource = resources.find((resourceObj) => resourceObj.id === id);
  return resource ? resource.label : id;
```